# ==============================================================================
# FIXING COMMIT DIFF
# ==============================================================================
# File: media/libcubeb/src/cubeb_wasapi.cpp
# Commit: 3ba614682b62
# Full Hash: 3ba614682b62df868e89da1d7705b3191954f0fc
# Author: Alex Chronopoulos <achronop@gmail.com>
# Date: 2019-04-25 16:18:07
# Description:
#   Bug 1545279 - Update cubeb from upstream to 3570749. r=padenot
#   
#   Pick commits:
#   3570749 - wasapi: tie monitor lifetime with notification client lifetime. BMO 1545279 (#505)
#   162625a - test: add option to get the posotion of a stream (#504)
# ==============================================================================

diff -r e1ab2cda0424 -r 3ba614682b62 media/libcubeb/src/cubeb_wasapi.cpp
--- a/media/libcubeb/src/cubeb_wasapi.cpp	Wed Apr 24 14:52:13 2019 +0000
+++ b/media/libcubeb/src/cubeb_wasapi.cpp	Thu Apr 25 09:22:35 2019 +0000
@@ -163,7 +163,6 @@
   /* Collection changed for output (render) devices. */
   cubeb_device_collection_changed_callback output_collection_changed_callback = nullptr;
   void * output_collection_changed_user_ptr = nullptr;
-  std::unique_ptr<monitor_device_notifications> monitor_notifications;
 };
 
 class wasapi_endpoint_notification_client;
@@ -308,6 +307,7 @@
 
   void notify(EDataFlow flow)
   {
+    XASSERT(cubeb_context);
     if (flow == eCapture && cubeb_context->input_collection_changed_callback) {
       bool res = SetEvent(input_changed);
       if (!res) {
@@ -450,6 +450,7 @@
   wasapi_collection_notification_client(cubeb * context)
     : ref_count(1)
     , cubeb_context(context)
+    , monitor_notifications(context)
   {
     XASSERT(cubeb_context);
   }
@@ -492,7 +493,7 @@
       if (FAILED(hr)) {
         return hr;
       }
-      cubeb_context->monitor_notifications->notify(flow);
+      monitor_notifications.notify(flow);
     }
     return S_OK;
   }
@@ -530,6 +531,7 @@
   LONG ref_count;
 
   cubeb * cubeb_context = nullptr;
+  monitor_device_notifications monitor_notifications;
 };
 
 class wasapi_endpoint_notification_client : public IMMNotificationClient
@@ -1336,8 +1338,6 @@
     context->device_collection_enumerator.reset();
   }
 
-  context->monitor_notifications.reset(new monitor_device_notifications(context));
-
   return hr;
 }
 
@@ -1352,7 +1352,6 @@
   context->collection_notification_client = nullptr;
   context->device_collection_enumerator = nullptr;
 
-  context->monitor_notifications.reset();
   return hr;
 }
 
